source/fuzz/transformation_flatten_conditional_branch.cpp

```diff
@@ -0,0 +1,595 @@
+
+      block = ir_context->cfg()->block(block_id);
+      block_id = block->terminator()->GetSingleWordInOperand(0);
+
+      // Find all the problematic instructions in the block (OpStore, OpLoad,
+      // OpFunctionCall).
+      std::vector<opt::Instruction*> problematic_instructions;
+
+      block->ForEachInst(
+          [&problematic_instructions](opt::Instruction* instruction) {
+            switch (instruction->opcode()) {
+              case SpvOpStore:
+              case SpvOpLoad:
+              case SpvOpFunctionCall:
+                problematic_instructions.push_back(instruction);
+                break;
+              default:
+                break;
+            }
+          });
+
+      uint32_t condition_id =
+          header_block->terminator()->GetSingleWordInOperand(0);
+
+      // Enclose all of the problematic instructions in conditionals, with the
+      // same condition as the selection construct being flattened.
+      for (auto instruction : problematic_instructions) {
+        // Collect the fresh ids needed by this instructions
+        uint32_t ids_needed = NumOfFreshIdsNeededByInstruction(instruction);
+        std::vector<uint32_t> fresh_ids;
+
+        // Get them from the map.
+        if (instructions_to_fresh_ids.count(instruction) != 0) {
+          fresh_ids = instructions_to_fresh_ids[instruction];
+        }
+
+        // Get the ones still needed from the overflow ids.
+        for (int still_needed = ids_needed - static_cast<int>(fresh_ids.size());
+             still_needed > 0; still_needed--) {
+          fresh_ids.push_back(message_.overflow_id(overflow_ids_used++));
+        }
+
+        // Enclose the instruction in a conditional and get the merge block
+        // generated by this operation (this is where all the next instructions
+        // will be).
+        block = EncloseInstructionInConditional(
+            ir_context, transformation_context, block, instruction, fresh_ids,
+            condition_id, branch == 1);
+      }
+
+      // If the next block is the convergence block and this is the true branch,
+      // record this as the last block in the true branch.
+      if (block_id == convergence_block_id && branch == 1) {
+        last_true_block = block;
+      }
+    }
+  }
+
+  // Get the condition operand and the ids of the first blocks of the true and
+  // false branches.
+  auto condition_operand = branch_instruction->GetInOperand(0);
+  uint32_t first_true_block_id = branch_instruction->GetSingleWordInOperand(1);
+  uint32_t first_false_block_id = branch_instruction->GetSingleWordInOperand(2);
+
+  // The current header should unconditionally branch to the first block in the
+  // true branch, if there exists a true branch, and to the first block in the
+  // false branch if there is no true branch.
+  uint32_t after_header = first_true_block_id != convergence_block_id
+                              ? first_true_block_id
+                              : first_false_block_id;
```

Comment:
SPIR-V has a rule that all blocks should appear syntactically before any blocks they dominate. You've inserted all blocks from the false branch right before the blocks from the true branch. Thus, I think should branch unconditionally to the first false block if there is one or to the first true block otherwise. @stefanomil, what do you think?

Reply:
Or you could've inserted the blocks from the true branch first.

Reply:
I have inserted all blocks from the true branch right before the false branch (I first add the false branch right after the header and then the true branch right after the header). So it should branch unconditionally to the true branch, if it exists

Reply:
(The comment saying otherwise was wrong)